fonduer/supervision/lf_helpers/visual.py

```diff
@@ -0,0 +1,433 @@
+        ):
+            yield ngram
+
+
+def get_vert_ngrams(
+    span, attrib="words", n_min=1, n_max=1, lower=True, from_sentence=True
+):
+    """Return all ngrams which are visually vertivally aligned with the Span.
+
+    Note that if a candidate is passed in, all of its Spans will be searched.
+
+    :param span: The Span to evaluate
+    :param attrib: The token attribute type (e.g. words, lemmas, poses)
+    :param n_min: The minimum n of the ngrams that should be returned
+    :param n_max: The maximum n of the ngrams that should be returned
+    :param lower: If True, all ngrams will be returned in lower case
+    :param from_sentence: If True, returns ngrams from any horizontally aligned
+        Sentences, rather than just horizontally aligned ngrams themselves.
+    :rtype: a *generator* of ngrams
+    """
+    spans = [span] if isinstance(span, TemporarySpan) else span.get_contexts()
+    for span in spans:
+        for ngram in _get_direction_ngrams(
+            "vert", span, attrib, n_min, n_max, lower, from_sentence
+        ):
+            yield ngram
+
+
+def _get_direction_ngrams(direction, c, attrib, n_min, n_max, lower, from_sentence):
+    # TODO: this currently looks only in current table;
+    #   precompute over the whole document/page instead
+    bbox_direction_aligned = (
+        bbox_vert_aligned if direction == "vert" else bbox_horz_aligned
+    )
+    ngrams_space = Ngrams(n_max=n_max, split_tokens=[])
+    f = (lambda w: w.lower()) if lower else (lambda w: w)
+    spans = [c] if isinstance(c, TemporarySpan) else c.get_contexts()
+    for span in spans:
+        if not span.sentence.is_tabular() or not span.sentence.is_visual():
+            continue
+        for sentence in span.sentence.table.sentences:
+            if from_sentence:
+                if (
+                    bbox_direction_aligned(
+                        bbox_from_sentence(sentence), bbox_from_span(span)
+                    )
+                    and sentence is not span.sentence
+                ):
+                    for ngram in tokens_to_ngrams(
+                        getattr(sentence, attrib), n_min=n_min, n_max=n_max, lower=lower
+                    ):
+                        yield ngram
+            else:
+                for ts in ngrams_space.apply(sentence):
+                    if bbox_direction_aligned(
+                        bbox_from_span(ts), bbox_from_span(span)
+                    ) and not (
+                        sentence == span.sentence and ts.get_span() in span.get_span()
+                    ):
+                        yield f(ts.get_span())
+
+
+def get_vert_ngrams_left(c):
+    """Not implemented."""
+    # TODO
+    return
+
+
+def get_vert_ngrams_right(c):
+    """Not implemented."""
+    # TODO
+    return
+
+
+def get_vert_ngrams_center(c):
+    """Not implemented."""
+    # TODO
+    return
+
+
+def get_visual_header_ngrams(c, axis=None):
```

Comment:
I don't expose these unimplemented helpers in the __init__.py

Reply:
👍